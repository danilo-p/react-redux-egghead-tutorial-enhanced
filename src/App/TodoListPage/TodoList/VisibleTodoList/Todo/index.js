import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'semantic-ui-react';
import styles from './index.styl';

/**
 * Describes a todo item.
 *
 * @param {Object} props The Todo's component properties
 * @param {Object} props.onClick The todo's click handler
 * @param {Object} props.completed The todo's completeness
 * @param {Object} props.text The todo's content
 * @returns {jsx} The Todo's component template
 */
const Todo = ({
  onClick,
  completed,
  text,
}) => (
  <li>
    <Checkbox
      className={styles.Todo__Checkbox}
      onClick={onClick}
      label={{ children: text }}
      checked={completed}
    />
  </li>
);

Todo.propTypes = {
  onClick: PropTypes.func.isRequired,
  completed: PropTypes.bool.isRequired,
  text: PropTypes.string.isRequired,
};

export default Todo;
