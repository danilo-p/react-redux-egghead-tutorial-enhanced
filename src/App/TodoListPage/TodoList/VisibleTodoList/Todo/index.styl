.Todo__Checkbox:global(.checked) label
  text-decoration: line-through;