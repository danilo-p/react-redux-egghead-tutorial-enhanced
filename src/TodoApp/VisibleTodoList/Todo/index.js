import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'semantic-ui-react';

import styles from './index.styl';

const Todo = ({
  onClick,
  completed,
  text,
}) => (
  <li>
    <Checkbox
      className={styles.Todo__Checkbox}
      onClick={onClick}
      label={{ children: text }}
      checked={completed}
    />
  </li>
);

Todo.propTypes = {
  onClick: PropTypes.func.isRequired,
  completed: PropTypes.bool.isRequired,
  text: PropTypes.string.isRequired,
};

export default Todo;
