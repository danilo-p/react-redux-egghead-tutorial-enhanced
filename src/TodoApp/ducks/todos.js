import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [
    {
      id: 0,
      text: 'test0',
      completed: false,
    },
    {
      id: 1,
      text: 'test1',
      completed: false,
    },
    {
      id: 2,
      text: 'test2',
      completed: false,
    },
  ],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
