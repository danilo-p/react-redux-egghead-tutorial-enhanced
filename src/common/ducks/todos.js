import todo, {
  ADD,
  TOGGLE,
} from './todo';

// Actions
export const REMOVE = 'REMOVE_TODO';
export { ADD, TOGGLE };

// Reducer
export default (
  state = [],
  action,
) => {
  switch (action.type) {
  case ADD:
    return [
      ...state,
      todo(undefined, action),
    ];
  case TOGGLE:
    return state.map(t => todo(t, action));
  case REMOVE:
    return []; // TODO: implement the remove todo
  default:
    return state;
  }
};
