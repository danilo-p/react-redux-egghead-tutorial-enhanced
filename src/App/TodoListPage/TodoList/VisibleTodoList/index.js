import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { toggleTodo } from 'common/ducks/todos';
import Todo from './Todo/index';

/**
 * Describes the todo list
 *
 * @param {any} props The component properties
 * @param {Array} props.todosList The array containing all the todos.
 * @param {Function} props.onTodoClick The todo click handler
 * @returns {jsx} The component template
 */
const VisibleTodoList = ({
  todosList,
  onTodoClick,
}) => (
  <ul>
    {todosList.map(t => (
      <Todo
        key={t.id}
        {...t}
        onClick={() => onTodoClick(t.id)}
      />
    ))}
  </ul>
);

/**
 * Utility for filtering an array of todos given a filter.
 *
 * @param {Array} todosList The todos list
 * @param {String} filter The filter to be applied on the todosList array.
 * @returns {jsx} The filtered array
 */
const getVisibleTodos = (
  todosList,
  filter,
) => {
  switch (filter) {
  case 'SHOW_COMPLETED':
    return todosList.filter(t => t.completed);
  case 'SHOW_ACTIVE':
    return todosList.filter(t => !t.completed);
  case 'SHOW_ALL':
  default:
    return todosList;
  }
};

VisibleTodoList.propTypes = {
  todosList: PropTypes.arrayOf(PropTypes.object).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  todosList: getVisibleTodos(
    state.todos,
    state.visibilityFilter,
  ),
});

const mapDispatchToProps = dispatch => ({
  onTodoClick: id => ({
    dispatch: dispatch(toggleTodo(id)),
  }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(VisibleTodoList);
