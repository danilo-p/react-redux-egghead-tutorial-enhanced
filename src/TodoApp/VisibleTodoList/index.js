import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import Todo from './Todo';
import { toggleTodo } from '../ducks/todo';

const getVisibleTodos = (
  todosList,
  filter,
) => {
  switch (filter) {
  case 'SHOW_COMPLETED':
    return todosList.filter(t => t.completed);
  case 'SHOW_ACTIVE':
    return todosList.filter(t => !t.completed);
  case 'SHOW_ALL':
  default:
    return todosList;
  }
};

const TodoList = ({
  todosList,
  onTodoClick,
}) => (
  <ul>
    {todosList.map(t => (
      <Todo
        key={t.id}
        {...t}
        onClick={() => onTodoClick(t.id)}
      />
    ))}
  </ul>
);

TodoList.propTypes = {
  todosList: PropTypes.arrayOf(PropTypes.object).isRequired,
  onTodoClick: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  todosList: getVisibleTodos(
    state.todos,
    state.visibilityFilter,
  ),
});

const mapDispatchToProps = dispatch => ({
  onTodoClick: id => ({
    dispatch: dispatch(toggleTodo(id)),
  }),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(TodoList);
