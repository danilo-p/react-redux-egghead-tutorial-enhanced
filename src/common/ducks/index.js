/**
 * Module that bundles all the ducks in one reducer. It imports all the ducks
 * and combine their reducers with the combineReducers method of the redux
 * library.
 *
 * @module common/ducks
 */

import { combineReducers } from 'redux';

import todo from './todo';
import todos from './todos';
import visibilityFilter from './visibilityFilter';

// Here all the reducers are combined into a single one reducer that is
// exported as default and used to create the application store.
export default combineReducers({
  todo,
  todos,
  visibilityFilter,
});
