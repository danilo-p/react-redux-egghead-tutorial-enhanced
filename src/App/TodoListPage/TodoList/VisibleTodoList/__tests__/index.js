import React from 'react';
import { INITIAL_STATE } from 'common/ducks';
import { createTodoId, TOGGLE } from 'common/ducks/todos';
import configureMockStore from 'redux-mock-store';
import VisibleTodoList, {
  getVisibleTodos,
  mapDispatchToProps,
} from '../';
import Todo from '../Todo';

const mockStore = configureMockStore();

describe('VisibleTodoList', () => {
  let visibleTodoList = null;
  let store = null;
  const todosList = [];

  for (let i = 1; i <= 5; i += 1) {
    const text = `Todo ${i}`;
    todosList.push({
      id: createTodoId(text),
      text,
      completed: i % 2 === 0,
    });
  }

  Object.freeze(todosList);

  beforeEach(() => {
    store = mockStore(Object.assign(INITIAL_STATE, {
      todos: todosList,
      filter: 'SHOW_ALL',
    }));

    visibleTodoList = mount(<VisibleTodoList store={store} />);
  });

  it('should render the component', () => {
    expect(visibleTodoList).toMatchSnapshot();
  });

  it('always render one "ul" element with everything else', () => {
    expect(visibleTodoList.find('ul').length).toBeGreaterThan(0);
  });

  it('always render one "Todo" element for each todo on the state', () => {
    expect(visibleTodoList.find(Todo).length)
      .toBe(store.getState().todos.length);
  });

  it(
    'should dispatch an TOGGLE_TODO action on a Todo click with the todo id',
    () => {
      const todoId = todosList[0].id;

      const todo = visibleTodoList
        .find('ul')
        .findWhere(node => node.props().id === todoId)
        .first();

      console.log(todo.props());

      todo.simulate('click');

      setTimeout(() => { // Wait for state update
        const actions = store.getActions();

        expect(actions.length).toBe(1);

        const action = actions[0];

        expect(action.type).toBe(TOGGLE);
        expect(action.id).toBe(todoId);
      }, 1000);
    },
  );

  describe('getVisibleTodos', () => {
    it(
      'should return only completed items for the SHOW_COMPLETED filter',
      () => {
        const filteredArray = getVisibleTodos(todosList, 'SHOW_COMPLETED');
        const completedCount =
          filteredArray.reduce((count, todo) => (
            todo.completed
              ? count + 1
              : count
          ), 0);
        expect(completedCount).toBe(filteredArray.length);
      },
    );

    it('should return only active items for the SHOW_ACTIVE filter', () => {
      const filteredArray = getVisibleTodos(todosList, 'SHOW_ACTIVE');
      const activeCount =
          filteredArray.reduce((count, todo) => (
            !todo.completed
              ? count + 1
              : count
          ), 0);
      expect(activeCount).toBe(filteredArray.length);
    });

    it('should return all items for the SHOW_ALL and any other filter', () => {
      let filteredArray = [];

      filteredArray = getVisibleTodos(todosList, 'SHOW_ALL');
      expect(filteredArray.length).toBe(todosList.length);

      filteredArray = getVisibleTodos(todosList, 'ANY_OTHER_FILTER');
      expect(filteredArray.length).toBe(todosList.length);
    });
  });

  describe('mapDispatchToProps', () => {
    it('should dispatch an TOGGLE_TODO action with the given id', () => {
      const todoText = 'Todo text';
      const todoId = createTodoId(todoText);

      mapDispatchToProps(store.dispatch).onTodoClick(todoId);

      const actions = store.getActions();

      expect(actions.length).toBe(1);

      const action = actions[0];

      expect(action.type).toBe(TOGGLE);
      expect(action.id).toBe(todoId);
    });
  });
});
