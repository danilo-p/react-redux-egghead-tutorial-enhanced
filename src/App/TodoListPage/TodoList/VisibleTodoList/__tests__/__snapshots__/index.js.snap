// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VisibleTodoList should render the component 1`] = `
<Connect(VisibleTodoList)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <VisibleTodoList
    onTodoClick={[Function]}
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    todosList={
      Array [
        Object {
          "completed": false,
          "id": "1515032510051Todo 1",
          "text": "Todo 1",
        },
        Object {
          "completed": true,
          "id": "1515032510051Todo 2",
          "text": "Todo 2",
        },
        Object {
          "completed": false,
          "id": "1515032510051Todo 3",
          "text": "Todo 3",
        },
        Object {
          "completed": true,
          "id": "1515032510051Todo 4",
          "text": "Todo 4",
        },
        Object {
          "completed": false,
          "id": "1515032510051Todo 5",
          "text": "Todo 5",
        },
      ]
    }
  >
    <ul>
      <Todo
        completed={false}
        id="1515032510051Todo 1"
        key="1515032510051Todo 1"
        onClick={[Function]}
        text="Todo 1"
      >
        <li>
          <Checkbox
            checked={false}
            className="Todo__Checkbox"
            label={
              Object {
                "children": "Todo 1",
              }
            }
            onClick={[Function]}
            type="checkbox"
          >
            <div
              className="ui checkbox Todo__Checkbox"
              onChange={[Function]}
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <input
                checked={false}
                className="hidden"
                name={undefined}
                readOnly={true}
                tabIndex={0}
                type="checkbox"
                value={undefined}
              />
              <label
                htmlFor={undefined}
              >
                Todo 1
              </label>
            </div>
          </Checkbox>
        </li>
      </Todo>
      <Todo
        completed={true}
        id="1515032510051Todo 2"
        key="1515032510051Todo 2"
        onClick={[Function]}
        text="Todo 2"
      >
        <li>
          <Checkbox
            checked={true}
            className="Todo__Checkbox"
            label={
              Object {
                "children": "Todo 2",
              }
            }
            onClick={[Function]}
            type="checkbox"
          >
            <div
              className="ui checked checkbox Todo__Checkbox"
              onChange={[Function]}
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <input
                checked={true}
                className="hidden"
                name={undefined}
                readOnly={true}
                tabIndex={0}
                type="checkbox"
                value={undefined}
              />
              <label
                htmlFor={undefined}
              >
                Todo 2
              </label>
            </div>
          </Checkbox>
        </li>
      </Todo>
      <Todo
        completed={false}
        id="1515032510051Todo 3"
        key="1515032510051Todo 3"
        onClick={[Function]}
        text="Todo 3"
      >
        <li>
          <Checkbox
            checked={false}
            className="Todo__Checkbox"
            label={
              Object {
                "children": "Todo 3",
              }
            }
            onClick={[Function]}
            type="checkbox"
          >
            <div
              className="ui checkbox Todo__Checkbox"
              onChange={[Function]}
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <input
                checked={false}
                className="hidden"
                name={undefined}
                readOnly={true}
                tabIndex={0}
                type="checkbox"
                value={undefined}
              />
              <label
                htmlFor={undefined}
              >
                Todo 3
              </label>
            </div>
          </Checkbox>
        </li>
      </Todo>
      <Todo
        completed={true}
        id="1515032510051Todo 4"
        key="1515032510051Todo 4"
        onClick={[Function]}
        text="Todo 4"
      >
        <li>
          <Checkbox
            checked={true}
            className="Todo__Checkbox"
            label={
              Object {
                "children": "Todo 4",
              }
            }
            onClick={[Function]}
            type="checkbox"
          >
            <div
              className="ui checked checkbox Todo__Checkbox"
              onChange={[Function]}
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <input
                checked={true}
                className="hidden"
                name={undefined}
                readOnly={true}
                tabIndex={0}
                type="checkbox"
                value={undefined}
              />
              <label
                htmlFor={undefined}
              >
                Todo 4
              </label>
            </div>
          </Checkbox>
        </li>
      </Todo>
      <Todo
        completed={false}
        id="1515032510051Todo 5"
        key="1515032510051Todo 5"
        onClick={[Function]}
        text="Todo 5"
      >
        <li>
          <Checkbox
            checked={false}
            className="Todo__Checkbox"
            label={
              Object {
                "children": "Todo 5",
              }
            }
            onClick={[Function]}
            type="checkbox"
          >
            <div
              className="ui checkbox Todo__Checkbox"
              onChange={[Function]}
              onClick={[Function]}
              onMouseDown={[Function]}
            >
              <input
                checked={false}
                className="hidden"
                name={undefined}
                readOnly={true}
                tabIndex={0}
                type="checkbox"
                value={undefined}
              />
              <label
                htmlFor={undefined}
              >
                Todo 5
              </label>
            </div>
          </Checkbox>
        </li>
      </Todo>
    </ul>
  </VisibleTodoList>
</Connect(VisibleTodoList)>
`;
